lib/collections.go

```diff
@@ -949,6 +1022,75 @@ func zipLists(arg0, arg1 ref.Val) ref.Val {
 	return types.NewRefValMap(types.DefaultTypeAdapter, m)
 }
 
+func mapKeys(val ref.Val) ref.Val {
+	mapK, ok := val.(traits.Mapper)
+	if !ok {
+		return types.ValOrErr(mapK, "no such overload")
+	}
+	n, ok := mapK.Size().(types.Int)
+	if !ok {
+		return types.NewErr("unable to get size of map")
+	}
+	keys := make([]ref.Val, 0, n)
+	if mapK.Size() != types.IntZero {
+		canSort := true
+		it := mapK.Iterator()
+		for it.HasNext() == types.True {
+			k := it.Next()
+			keys = append(keys, k)
+			_, ok := k.(traits.Comparer)
+			if !ok {
+				canSort = false
+			}
+		}
+		if canSort {
+			sort.Slice(keys, func(i, j int) bool {
+				return keys[i].(traits.Comparer).Compare(keys[j]) == types.Int(-1)
+			})
+		}
+	}
+	return types.NewRefValList(types.DefaultTypeAdapter, keys)
+}
+
+func mapValues(val ref.Val) ref.Val {
+	mapK, ok := val.(traits.Mapper)
+	if !ok {
+		return types.ValOrErr(mapK, "no such overload")
+	}
+	n, ok := mapK.Size().(types.Int)
+	if !ok {
+		return types.NewErr("unable to get size of map")
+	}
+	values := make([]ref.Val, 0, n)
+	type kv struct {
+		Key   ref.Val
+		Value ref.Val
+	}
+	if mapK.Size() != types.IntZero {
+		canSort := true
+		it := mapK.Iterator()
+		ss := make([]kv, 0, n)
+		for it.HasNext() == types.True {
+			k := it.Next()
+			v := mapK.Get(k)
+			ss = append(ss, kv{k, v})
+			_, ok := k.(traits.Comparer)
+			if !ok {
+				canSort = false
+			}
+		}
+		if canSort {
+			sort.Slice(ss, func(i, j int) bool {
+				return ss[i].Key.(traits.Comparer).Compare(ss[j].Key) == types.Int(-1)
+			})
+		}
```

Comment:
```suggestion
	type valComparer interface {
		ref.Val
		traits.Comparer
	}
	type kv struct {
		Key   valComparer
		Value ref.Val
	}
	if mapK.Size() != types.IntZero {
		canSort := true
		it := mapK.Iterator()
		ss := make([]kv, 0, n)
		for it.HasNext() == types.True {
			k := it.Next()
			v := mapK.Get(k)
			ck, ok := k.(valComparer)
			if !ok {
				canSort = false
			}
			ss = append(ss, kv{ck, v})
		}
		if canSort {
			sort.Slice(ss, func(i, j int) bool {
				return ss[i].Key.Compare(ss[j].Key) == types.Int(-1)
			})
		}
```

Reply:
Great, thanks for the recommendation here. Updated